packages/yew-macro/src/hook/body.rs

```diff
@@ -0,0 +1,118 @@
+use proc_macro2::Span;
+use proc_macro_error::emit_error;
+use std::sync::{Arc, Mutex};
+use syn::spanned::Spanned;
+use syn::visit_mut::VisitMut;
+use syn::{
+    parse_quote_spanned, visit_mut, Expr, ExprCall, ExprClosure, ExprForLoop, ExprIf, ExprLoop,
+    ExprMatch, ExprWhile, Ident, Item,
+};
+
+#[derive(Debug, Default)]
+pub struct BodyRewriter {
+    branch_lock: Arc<Mutex<()>>,
+}
+
+impl BodyRewriter {
+    fn is_branched(&self) -> bool {
+        self.branch_lock.try_lock().is_err()
+    }
+
+    fn with_branch<F, O>(&mut self, f: F) -> O
+    where
+        F: FnOnce(&mut BodyRewriter) -> O,
+    {
+        let branch_lock = self.branch_lock.clone();
+        let _branched = branch_lock.try_lock();
+        f(self)
+    }
+}
+
+impl VisitMut for BodyRewriter {
+    fn visit_expr_call_mut(&mut self, i: &mut ExprCall) {
+        let ctx_ident = Ident::new("ctx", Span::mixed_site());
+
+        // Only rewrite hook calls.
+        if let Expr::Path(ref m) = &*i.func {
+            if let Some(m) = m.path.segments.last().as_ref().map(|m| &m.ident) {
+                if m.to_string().starts_with("use_") {
+                    if self.is_branched() {
+                        emit_error!(m, "hooks cannot be called at this position.");
+                    } else {
+                        *i = parse_quote_spanned! { i.span() => ::yew::functional::Hook::run(#i, #ctx_ident) };
+                    }
+
+                    return;
+                }
```

Comment:
I take it that this design means I can't all any function starting with `use_` that is not a hook and false positives are a conscious trade-off? How do I turn this off when such a false positive occurs?
I have some doubts that this kind of syntax guided rewriting is maintainable in the long term, do you know any bigger library that uses this approach successfully to convince me that it's not going to break on us in the future?

Reply:
> How do I turn this off when such a false positive occurs?

```
use some_crate::{use_something as not_a_hook} ;
```

> that uses this approach successfully

This is used in std as well, which I will reply to the main comment instead.

Reply:
@futursolo we should mention this in the error to be more user friendly

Reply:
There's no way to do it. The procedural does not know the type of your expression.

The only error will be from the compiler saying that `use_something` does not implement `Hook`.